! pip install scikit-learn

In [37]:
from sklearn.datasets import load_iris 
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split 

In [38]:
import pandas as pd 

#붓꽃 데이터 셋 로딩
iris = load_iris()
print(type(iris))
keys = iris.keys()
print("붓꽃 데이터 세트의 키들 : ", keys)
print("----------------------------------------")

#iris data는 iris 데이터 셋에서 피처(feature)만으로 된 데이터를 numpy로 가지고 있음
iris_data = iris.data 

#iris.target 은 붓꽃 데이터 셋에서 레이블(결정값) 데이터를 numpy로 가지고 있음
iris_label = iris.target
print("iris target 값 : ", iris_label)
print("iris target 명 : ", iris.target_names)
print("----------------------------------------")

#붓꽃 데이터 셋을 자세히 보기 위해 DataFrame으로 변환함
iris_df = pd.DataFrame(data = iris_data, columns = iris.feature_names)
iris_df['label'] = iris.target
iris_df.head(3)


<class 'sklearn.utils.Bunch'>
붓꽃 데이터 세트의 키들 :  dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename'])
----------------------------------------
iris target 값 :  [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]
iris target 명 :  ['setosa' 'versicolor' 'virginica']
----------------------------------------


sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   

   label  
0      0  
1      0  
2      0

In [39]:
x_train, x_test, y_train, y_test = train_test_split(iris_data, iris_label,test_size = 0.2, random_state =11)


In [40]:
#DecisionTreeClassifier 객체 생성 
dt_clf = DecisionTreeClassifier(random_state = 11)

# 학습 데이터를 이용하여 학습 수행 
dt_clf.fit(x_train, y_train)

DecisionTreeClassifier(random_state=11)

In [41]:
#학습이 완료된 DecisionTreeClassifier 객체에서 테스트 데이터 세트로 예측 수행
pred = dt_clf.predict(x_test)

In [45]:
from sklearn.metrics import accuracy_score

print("예측 정확도 : {0:.4f}".format(accuracy_score(y_test,pred)))

예측 정확도 : 0.9333


지도학습 : fit - predict   
-1. 분류 - DecisionTreeClassifier,  RandomForestClassifier, GradientBoostingClassifier, GaussianNB, SVM  

-2. 회귀(예측) - LinearRegression, Ridge, Lasso, RandomForestRegressor, GradientBoosting, GradientBoostingRegressor 

  
  

비지도학습 : fit - transform 

In [46]:
print("feature_names 의 type : ", type(iris.feature_names))
print("feature_names 의 길이 : ", len(iris.feature_names))
print(iris.feature_names)
print("----------------------------------------")
print("target_names 의 type : ", type(iris.target_names))
print("target_names 의 길이 : ", len(iris.target_names))
print("target_names 의 shape : ", iris.target_names.shape)
print(iris.target_names)
print("----------------------------------------")
print("data의 type : ", type(iris.data))
print("data의 shape : ", iris.data.shape)
print(iris['data'])
print("----------------------------------------")
print("target 의 type : ", type(iris.target))
print("target 의 shape : ", iris.target.shape)
print(iris.target)

feature_names 의 type :  <class 'list'>
feature_names 의 길이 :  4
['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']
----------------------------------------
target_names 의 type :  <class 'numpy.ndarray'>
target_names 의 길이 :  3
target_names 의 shape :  (3,)
['setosa' 'versicolor' 'virginica']
----------------------------------------
data의 type :  <class 'numpy.ndarray'>
data의 shape :  (150, 4)
[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]
 [5.4 3.7 1.5 0.2]
 [4.8 3.4 1.6 0.2]
 [4.8 3.  1.4 0.1]
 [4.3 3.  1.1 0.1]
 [5.8 4.  1.2 0.2]
 [5.7 4.4 1.5 0.4]
 [5.4 3.9 1.3 0.4]
 [5.1 3.5 1.4 0.3]
 [5.7 3.8 1.7 0.3]
 [5.1 3.8 1.5 0.3]
 [5.4 3.4 1.7 0.2]
 [5.1 3.7 1.5 0.4]
 [4.6 3.6 1.  0.2]
 [5.1 3.3 1.7 0.5]
 [4.8 3.4 1.9 0.2]
 [5.  3.  1.6 0.2]
 [5.  3.4 1.6 0.4]
 [5.2 3.5 1.5 0.2]
 [5.2 3.4 1.4 0.2]
 [4.7 3.2 1.6 0.2]
 [4.8

학습데이터 셋 : 함수를 형성하기 위해 필요한 dataset


In [25]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.datasets import load_iris 
from sklearn.model_selection import train_test_split 

dt_clf = DecisionTreeClassifier()
iris_data = load_iris()

x_train, x_test, y_train, y_test = train_test_split(
    iris_data.data, iris_data.target,test_size = 0.3, random_state=121)

dt_clf.fit(x_train, y_train)
pred = dt_clf.predict(x_test)
print("예측 정확도 : {0:.4f}".format(accuracy_score(y_test,pred)))

예측 정확도 : 0.9556


K폴드 교차 검증  

-일반 K폴드 : 앞에서 부터 차례대로 데이터를 사용하면 됨 

-Stratified K폴드 : 학습데이터와 검증 데이터 set 이 가지는 레이블 분포도가 유사하도록 검증 데이터 추출 (사용처 ex. 1 ~ 50까지는 setosa , 50~100 versicolor , 150까지는 virginica 인 것은 적절히 각 y값(레이블)에서 적절히 추출해 와야 함)